### PR TITLE
[Enhancement] Adjust downgrade strategy of chunks_partitioner

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Enhancement

If we find the partition cardinality is high, then we will downgrade `chunks_partitioner` to streaming mode. But it only happens when we've done processing the entire chunk. 

Think about the situation that each partition has only one row, then the first comming chunk contains 4096 partitions, we need to first process the whole chunk, then we can decide whether to downgrade or not. And this downgrade procedure can happen more earlier.

So here is the idea, we can trigger downgrade process when we are processing the chunk, if we find the partition num is reaching the threshold, we can immediately stop the current process and switch to streaming mode.

### CODE REVIEW

File: be/src/exec/vectorized/partition/partition_hash_map.h [hunk 1]
```diff
@@ -92,11 +106,20 @@ struct PartitionHashMapBase {
...
+            return;
+        }
         phmap::flat_hash_set<typename HashMap::key_type, typename HashMap::hasher, typename HashMap::key_equal,
                              typename HashMap::allocator_type>
                 visited_keys(chunk->num_rows());
         const auto size = chunk->num_rows();
-        for (uint32_t i = 0; i < size; i++) {
+        uint32_t i = 0;
+        for (; i < size; i++) {
+            check_downgrade(hash_map);
```
Comment (mofeiatwork) [inline be/src/exec/vectorized/partition/partition_hash_map.h:118]: Would it hurt the performance? Check downgrade each row?

File: be/src/exec/vectorized/partition/partition_hash_map.h [hunk 2]
```diff
@@ -113,11 +136,19 @@ struct PartitionHashMapBase {
...
         }
 
         for (const auto& key : visited_keys) {
             flush(*(hash_map[key]), chunk);
         }
+
+        // The first i rows has been pushed into hash_map
+        if (is_downgrade && i > 0) {
+            for (auto& column : chunk->columns()) {
+                column->remove_first_n_values(i);
```
Comment (kangkaisen) [inline be/src/exec/vectorized/partition/partition_hash_map.h:149]: shouldn't remove too often。

**Comment by kangkaisen** [comment]
```
then we will downgrade chunks_partitioner to streaming mode. 
```
Why we call switch to streaming mode is `downgrade`?

**Review by Seaven — DISMISSED** [review]

File: be/src/exec/vectorized/partition/partition_hash_map.h (hunk 2 above)
Comment (liuyehcf) [reply be/src/exec/vectorized/partition/partition_hash_map.h:149]: It happens only once, i.e. when downgrade happens.

File: be/src/exec/vectorized/partition/partition_hash_map.h (hunk 1 above)
Comment (liuyehcf) [reply be/src/exec/vectorized/partition/partition_hash_map.h:118]: Good point of view, I'll see if I can optimize it

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by liuyehcf** [comment]
starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Pull request merged**